packages/cli/src/concurrency/concurrency-control.service.ts

```diff
@@ -0,0 +1,158 @@
+import { Logger } from '@/Logger';
+import config from '@/config';
+import { Service } from 'typedi';
+import { ConcurrencyQueue } from './concurrency-queue';
+import { UnknownExecutionModeError } from '@/errors/unknown-execution-mode.error';
+import { InvalidConcurrencyCapError } from '@/errors/invalid-concurrency-cap.error';
+import { ExecutionRepository } from '@/databases/repositories/execution.repository';
+import { License } from '@/License';
+import type { WorkflowExecuteMode as ExecutionMode } from 'n8n-workflow';
+
+@Service()
+export class ConcurrencyControlService {
+	private readonly isEnabled: boolean;
+
+	private readonly productionCap: number;
+
+	private readonly productionQueue: ConcurrencyQueue;
+
+	constructor(
+		private readonly logger: Logger,
+		private readonly executionRepository: ExecutionRepository,
+		private readonly license: License,
+	) {
+		this.productionCap = this.getProductionCap();
+
+		if (this.productionCap === 0) {
+			throw new InvalidConcurrencyCapError(this.productionCap);
+		}
+
+		if (this.productionCap < -1) {
+			this.productionCap = -1;
+		}
+
+		if (this.productionCap === -1 || config.getEnv('executions.mode') === 'queue') {
+			this.isEnabled = false;
+			this.log('Service disabled');
+			return;
+		}
+
+		this.productionQueue = new ConcurrencyQueue(this.productionCap);
+
+		this.logInit();
+
+		this.isEnabled = true;
+
+		this.productionQueue.on('execution-throttled', async (executionId: string) => {
+			this.log('Execution throttled', { executionId });
+		});
+
+		this.productionQueue.on('execution-released', async (executionId: string) => {
+			this.log('Execution released', { executionId });
+			await this.executionRepository.resetStartedAt(executionId);
+		});
+	}
+
+	/**
+	 * Block or let through an execution based on concurrency capacity.
+	 */
+	async check({ mode, executionId }: { mode: ExecutionMode; executionId: string }) {
+		if (!this.isEnabled || this.isUncapped(mode)) return;
+
+		await this.productionQueue.enqueue(executionId);
+	}
+
+	/**
+	 * Release capacity back so the next execution in the production queue can proceed.
+	 */
+	release({ mode }: { mode: ExecutionMode }) {
+		if (!this.isEnabled || this.isUncapped(mode)) return;
+
+		this.productionQueue.dequeue();
```

Comment:
would it be better if dequeue explicitly removed a specific executionId from the queue, instead of simply saying that "one slot just opened up" 🤔 ?

Reply:
Sorry I'm not seeing what you're seeing here - what is the benefit of this? `remove()` removes by execution ID but that's only needed when deleting an execution.